File: Classes/NumberEditTableCell.m
// NumberEditTableCell.m
//
// Kraftstoff


#import "NumberEditTableCell.h"
#import "FuelCalculatorController.h"


@implementation NumberEditTableCell

@synthesize textFieldSuffix;
@synthesize numberFormatter;
@synthesize alternateNumberFormatter;


- (void)finishConstruction
{
	[super finishConstruction];

    self.textField.keyboardType = UIKeyboardTypeNumberPad;
}


- (void)dealloc
{
    self.numberFormatter          = nil;
    self.alternateNumberFormatter = nil;
    self.textFieldSuffix          = nil;

    [super dealloc];
}


- (void)configureForData: (id)dataObject viewController: (id)viewController tableView: (UITableView*)tableView indexPath: (NSIndexPath*)indexPath
{
	[super configureForData: dataObject viewController: viewController tableView: tableView indexPath: indexPath];

    self.textFieldSuffix          = [(NSDictionary*)dataObject objectForKey: @"suffix"];
    self.numberFormatter          = [(NSDictionary*)dataObject objectForKey: @"formatter"];
    self.alternateNumberFormatter = [(NSDictionary*)dataObject objectForKey: @"alternateFormatter"];

    NSDecimalNumber *value = [self.delegate valueForIdentifier: self.valueIdentifier];

    if (value)
    {
        if (self.alternateNumberFormatter)
            self.textField.text = [self.alternateNumberFormatter stringFromNumber: value];
        else
            self.textField.text = [self.numberFormatter stringFromNumber: value];

        if (self.textFieldSuffix)
            self.textField.text = [self.textField.text stringByAppendingString: self.textFieldSuffix];
    }
    else
        self.textField.text = @"";
}



#pragma mark -
#pragma mark UITextFieldDelegate



// Implement special behavior for newly added characters
- (BOOL)textField: (UITextField*)aTextField shouldChangeCharactersInRange: (NSRange)range replacementString: (NSString*)string
{
    // Modify text
    NSString *text         = [aTextField text];
    NSDecimalNumber *value = (NSDecimalNumber*)[self.numberFormatter numberFromString: text];
    NSDecimalNumber *ten   = [NSDecimalNumber decimalNumberWithMantissa: 1 exponent: 1 isNegative: NO];
    NSDecimalNumber *scale = [NSDecimalNumber decimalNumberWithMantissa: 1 exponent: [self.numberFormatter maximumFractionDigits] isNegative: NO];

    if (range.length == 0)
    {
        if (range.location == [text length] && [string length] == 1)
        {
            // New character must be a digit
            NSDecimalNumber *digit = [NSDecimalNumber decimalNumberWithString: string];

            if ([digit isEqual: [NSDecimalNumber notANumber]])
                return NO;

            // Special shift semantics when appending at end of string
            value = [value decimalNumberByMultiplyingBy: ten];
            value = [value decimalNumberByAdding: [digit decimalNumberByDividingBy: scale]];
        }
        else
        {
            // Normal insert otherwise
            text  = [text stringByReplacingCharactersInRange: range withString: string];
            text  = [text stringByReplacingOccurrencesOfString: [self.numberFormatter groupingSeparator] withString: @""];
            value = (NSDecimalNumber*)[self.numberFormatter numberFromString: text];
        }

        // Don't append when the result gets too large or below zero
        if ([value compare: [NSDecimalNumber decimalNumberWithMantissa: 1 exponent: 6 isNegative: NO]] != NSOrderedAscending)
            return NO;

        if ([value compare: [NSDecimalNumber zero]] == NSOrderedAscending)
            return NO;
    }

    else if (range.location >= [text length] - 1)
    {
        NSDecimalNumberHandler *handler = [NSDecimalNumberHandler decimalNumberHandlerWithRoundingMode: NSRoundDown
                                                                                                 scale: (short)[self.numberFormatter maximumFractionDigits]
                                                                                      raiseOnExactness: NO
                                                                                       raiseOnOverflow: NO
                                                                                      raiseOnUnderflow: NO
                                                                                   raiseOnDivideByZero: NO];

        // Delete only the last digit
        value = [value decimalNumberByDividingBy: ten withBehavior: handler];
        value = [value decimalNumberByRoundingAccordingToBehavior:  handler];
    }

    [aTextField setText: [self.numberFormatter stringFromNumber: value]];

    // Tell delegate about new value
    [self.delegate valueChanged: value identifier: self.valueIdentifier];

    return NO;
}


// Reset to zero value on clear
- (BOOL)textFieldShouldClear: (UITextField*)aTextField
{
    NSNumber *clearedValue = [NSDecimalNumber zero];


    if (aTextField.editing)
    {
        aTextField.text = [self.numberFormatter stringFromNumber: clearedValue];
    }
    else
    {
        if (self.alternateNumberFormatter)
            aTextField.text = [self.alternateNumberFormatter stringFromNumber: clearedValue];
        else
            aTextField.text = [self.numberFormatter stringFromNumber: clearedValue];

        if (self.textFieldSuffix)
            aTextField.text = [aTextField.text stringByAppendingString: self.textFieldSuffix];
    }

    // Tell delegate about new value
    [self.delegate valueChanged: clearedValue identifier: self.valueIdentifier];

    return NO;
}


// Editing starts, remove suffix and switch to normal formatter
- (void)textFieldDidBeginEditing: (UITextField *)aTextField
{
    if (self.textFieldSuffix)
    {
        if ([aTextField.text hasSuffix: self.textFieldSuffix])
            aTextField.text = [aTextField.text substringToIndex: [aTextField.text length] - [self.textFieldSuffix length]];
    }

    if (self.alternateNumberFormatter)
    {
        NSDecimalNumber *value = (NSDecimalNumber*)[self.alternateNumberFormatter numberFromString: aTextField.text];

        if (value == nil)
            value = [NSDecimalNumber zero];

        aTextField.text = [self.numberFormatter stringFromNumber: value];
        [self.delegate valueChanged: value identifier: self.valueIdentifier];
    }
}


// Editing ends, switch back to alternate formatter and append specified suffix
- (void)textFieldDidEndEditing: (UITextField *)aTextField
{
    if (self.alternateNumberFormatter)
    {
        NSDecimalNumber *value = (NSDecimalNumber*)[self.numberFormatter numberFromString: aTextField.text];

        if (value == nil)
            value = [NSDecimalNumber zero];

        aTextField.text = [self.alternateNumberFormatter stringFromNumber: value];
        [self.delegate valueChanged: value identifier: self.valueIdentifier];
    }

    if (self.textFieldSuffix)
    {
        if (! [aTextField.text hasSuffix: self.textFieldSuffix])
            aTextField.text = [aTextField.text stringByAppendingString: self.textFieldSuffix];
    }

    [super textFieldDidEndEditing: aTextField];
}

@end

File: Classes/NumberEditTableCell.swift
//
//  NumberEditTableCell.swift
//  kraftstoff
//
//  Created by Ingmar Stein on 03.05.15.
//
//

import UIKit

final class NumberEditTableCell: EditablePageCell {
  var numberFormatter: NumberFormatter!
  var alternateNumberFormatter: NumberFormatter?
  var textFieldSuffix: String?

  required init() {
    super.init()

    textField.keyboardType = .numberPad
  }

  @available(*, unavailable)
  required init(coder _: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  private func updateTextFieldColorForValue(_ value: Any?) {
    let valid: Bool
    if let validator = delegate as? EditablePageCellValidator {
      valid = validator.valueValid(value, identifier: valueIdentifier)
    } else {
      valid = true
    }
    textField.textColor = valid ? .label : invalidTextColor
  }

  override func configureForData(_ dictionary: [String: Any], viewController: UIViewController, tableView: UITableView, indexPath: IndexPath) {
    super.configureForData(dictionary, viewController: viewController, tableView: tableView, indexPath: indexPath)

    textFieldSuffix = dictionary["suffix"] as? String
    numberFormatter = dictionary["formatter"] as? NumberFormatter
    alternateNumberFormatter = dictionary["alternateFormatter"] as? NumberFormatter

    let value = delegate.valueForIdentifier(valueIdentifier) as? Decimal
    if let value = value {
      textField.text = (alternateNumberFormatter ?? numberFormatter).string(from: value as NSNumber)

      if let suffix = textFieldSuffix {
        textField.text = textField.text!.appending(suffix)
      }
    } else {
      textField.text = ""
    }
    accessibilityIdentifier = valueIdentifier
    textField.accessibilityIdentifier = valueIdentifier

    updateTextFieldColorForValue(value)
  }

  // MARK: - UITextFieldDelegate

  // Implement special behavior for newly added characters
  @objc(textField:shouldChangeCharactersInRange:replacementString:)
  func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
    // Modify text
    var text = textField.text!
    guard let textValue = numberFormatter.number(from: text) as? Decimal else {
      return false
    }
    var value = textValue
    let scale = NSDecimalNumber(mantissa: 1, exponent: Int16(numberFormatter.maximumFractionDigits), isNegative: false) as Decimal

    if range.length == 0 {
      if range.location == text.count, string.count == 1 {
        // New character must be a digit
        guard let digit = Decimal(string: string) else { return false }

        if digit.isNaN {
          return false
        }

        // Special shift semantics when appending at end of string
        value = value << 1
        value += digit / scale
      } else {
        // Normal insert otherwise
        guard let editRange = Range(range, in: text) else { return false }
        text = text.replacingCharacters(in: editRange, with: string)
        text = text.replacingOccurrences(of: numberFormatter.groupingSeparator, with: "")
        guard let newValue = numberFormatter.number(from: text) as? Decimal else { return false }
        value = newValue
      }

      // Don't append when the result gets too large or below zero
      if value >= Decimal.fromLiteral(mantissa: 1, exponent: 6, isNegative: false) {
        return false
      }

      if value.isSignMinus {
        return false
      }
    } else if range.location >= text.count - 1 {
      let handler = NSDecimalNumberHandler(roundingMode: .down,
                                           scale: Int16(numberFormatter.maximumFractionDigits),
                                           raiseOnExactness: false,
                                           raiseOnOverflow: false,
                                           raiseOnUnderflow: false,
                                           raiseOnDivideByZero: false)

      // Delete only the last digit
      value = (value as NSDecimalNumber).multiplying(byPowerOf10: -1, withBehavior: handler).rounding(accordingToBehavior: handler) as Decimal
    }

    textField.text = numberFormatter.string(from: value as NSNumber)

    // Tell delegate about new value
    delegate.valueChanged(value, identifier: valueIdentifier)
    updateTextFieldColorForValue(value)

    return false
  }

  // Reset to zero value on clear
  func textFieldShouldClear(_ textField: UITextField) -> Bool {
    let clearedValue = Decimal(0)

    if textField.isEditing {
      textField.text = numberFormatter.string(from: clearedValue as NSNumber)
    } else {
      textField.text = (alternateNumberFormatter ?? numberFormatter).string(from: clearedValue as NSNumber)

      if let suffix = textFieldSuffix {
        textField.text = textField.text!.appending(suffix)
      }
    }

    // Tell delegate about new value
    delegate.valueChanged(clearedValue, identifier: valueIdentifier)
    updateTextFieldColorForValue(clearedValue)

    return false
  }

  // Editing starts, remove suffix and switch to normal formatter
  func textFieldDidBeginEditing(_ textField: UITextField) {
    if let suffix = textFieldSuffix {
      if textField.text!.hasSuffix(suffix) {
        textField.text = String(textField.text![..<textField.text!.index(textField.text!.endIndex, offsetBy: -suffix.count)])
      }
    }

    if let alternateNumberFormatter = alternateNumberFormatter {
      let value = alternateNumberFormatter.number(from: textField.text!) as? NSDecimalNumber ?? .zero
      textField.text = numberFormatter.string(from: value)
      delegate.valueChanged(value, identifier: valueIdentifier)
      updateTextFieldColorForValue(value)
    }
  }

  // Editing ends, switch back to alternate formatter and append specified suffix
  override func textFieldDidEndEditing(_ textField: UITextField, reason: UITextField.DidEndEditingReason) {
    super.textFieldDidEndEditing(textField, reason: reason)

    let value = numberFormatter.number(from: textField.text!) as? NSDecimalNumber ?? .zero

    if let alternateNumberFormatter = alternateNumberFormatter {
      textField.text = alternateNumberFormatter.string(from: value)
    }

    if let suffix = textFieldSuffix {
      if !textField.text!.hasSuffix(suffix) {
        textField.text = textField.text!.appending(suffix)
      }
    }

    delegate.valueChanged(value, identifier: valueIdentifier)
    updateTextFieldColorForValue(value)
  }
}
